### ISSUE TITLE
Throwing unexpected exceptions when using unicode in method signatures or class names

### ISSUE THREAD

**Issue by vinix128** [guillep/pharo-core#10805]
**Bug description**
When using unicode in method and class identifiers, utf-8 decoding errors often occur.

**To Reproduce**
The error often occurs when requesting different versions of a method.

**Expected behavior**
The Pharo environment should work correctly with utf-8 encoded files and unicode strings.

The error occurs because the read pointer in the change file is placed in the middle of a multibyte character, causing the decoder to throw an exception.

The problematic code is in the changeRecordsFrom:className:isMeta:do: method of the SourceFileArray class:

file position: (0 max: filePosition - 150). "Skip back to before the preamble"

**Screenshots**
![image](https://user-images.githubusercontent.com/98399922/150991267-15a3fb7a-f2c3-418f-8639-2f97aa41e8e8.png)

**Version information:**
 - OS: macOS
 - Version: 12.1
 - Pharo Version 5-10

**Comment by svenvc** [comment]
You are correct: you cannot move to an arbitrary byte level position in a variable length multibyte encoding like UTF8 and expect not to end up in the middle of a character, instead of between characters.

ZnUTF8Encoder>>#backOnStream: can move backwards from one legal position to the previous one.

Just like counting the characters as in ZnCharacterEncoder>>#encodedByteCountForCodePoints: this is necessarily a sequential loop.
```
| stream encoder |
stream := 'élève' utf8Encoded readStream.
stream setToEnd.
encoder := ZnUTF8Encoder default.
3 timesRepeat: [ encoder backOnStream: stream ].
(ZnCharacterReadStream on: stream encoding: encoder) next.
 "$è"
3 timesRepeat: [ encoder backOnStream: stream ].
(ZnCharacterReadStream on: stream encoding: encoder) next.
 "$é"
```
A check for being at the beginning needs to be added as well.

**Comment by Ducasse** [comment]
I do not think that method name and class identifier in Pharo support utf-8 and this is not high on our roadmap.

**Comment by svenvc** [comment]
> I do not think that method name and class identifier in Pharo support utf-8 and this is not high on our roadmap.

Maybe, but that is not the issue here, it is the use of non-ASCII in the method source code or comment, that we certainly support.

**Comment by svenvc** [comment]
SourceFile>>#getPreambleAt: was recently (2019) rewritten specifically for handling non-ASCII code

https://github.com/pharo-project/pharo/pull/2433

That might be useful, or could maybe be (re)used.

**Comment by Ducasse** [comment]
Ok thanks sven for the analysis and pointer.
